FtdiUsbSerialDxe console redirection support

filipleple opened the pull request.

filipleple committed:
FtdiUsbSerialDxe: First try RESET then PURGE

Some FTDI chips don't support RESET_PORT_PURGE_RX and require RESET_PORT_RESET
(0x0) instead.

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
FtdiUsbSerialDxe: ignore SerialIO Reset() failure

Some FTDI devices don't need resetting to function properly. For the
currently tested device, resetting caused errors. After ignoring them,
the device worked properly.

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

miczyg1 commented on Drivers/OptionRomPkg/Bus/Usb/FtdiUsbSerialDxe/FtdiUsbSerialDriver.c:
```diff
@@ -2070,6 +2136,22 @@ UsbSerialDriverBindingStart (
+  Status = gBS->InstallMultipleProtocolInterfaces (
+                    &UsbSerialDevice->ControllerHandle,
+                    &gEfiSimpleTextInProtocolGuid,
+                    &UsbSerialDevice->SerialIo,
```
You can't do that. SerialIO protocol has the following definition:

```C
struct _EFI_SERIAL_IO_PROTOCOL {
  ///
  /// The revision to which the EFI_SERIAL_IO_PROTOCOL adheres. All future revisions
  /// must be backwards compatible. If a future version is not backwards compatible,
  /// it is not the same GUID.
  ///
  UINT32                         Revision;
  EFI_SERIAL_RESET               Reset;
  EFI_SERIAL_SET_ATTRIBUTES      SetAttributes;
  EFI_SERIAL_SET_CONTROL_BITS    SetControl;
  EFI_SERIAL_GET_CONTROL_BITS    GetControl;
  EFI_SERIAL_WRITE               Write;
  EFI_SERIAL_READ                Read;
  ///
  /// Pointer to SERIAL_IO_MODE data.
  ///
  EFI_SERIAL_IO_MODE             *Mode;
  ///
  /// Pointer to a GUID identifying the device connected to the serial port.
  /// This field is NULL when the protocol is installed by the serial port
  /// driver and may be populated by a platform driver for a serial port
  /// with a known device attached. The field will remain NULL if there is
  /// no platform serial device identification information available.
  ///
  CONST EFI_GUID                 *DeviceTypeGuid; // Revision 1.1
};
```


Simple Text In and Out have the following definitions:

```C
///
/// The EFI_SIMPLE_TEXT_INPUT_PROTOCOL is used on the ConsoleIn device.
/// It is the minimum required protocol for ConsoleIn.
///
struct _EFI_SIMPLE_TEXT_INPUT_PROTOCOL {
  EFI_INPUT_RESET       Reset;
  EFI_INPUT_READ_KEY    ReadKeyStroke;
  ///
  /// Event to use with WaitForEvent() to wait for a key to be available
  ///
  EFI_EVENT             WaitForKey;
};

struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL {
  EFI_INPUT_RESET_EX                 Reset;
  EFI_INPUT_READ_KEY_EX              ReadKeyStrokeEx;
  ///
  /// Event to use with WaitForEvent() to wait for a key to be available.
  ///
  EFI_EVENT                          WaitForKeyEx;
  EFI_SET_STATE                      SetState;
  EFI_REGISTER_KEYSTROKE_NOTIFY      RegisterKeyNotify;
  EFI_UNREGISTER_KEYSTROKE_NOTIFY    UnregisterKeyNotify;
};


struct _EFI_SIMPLE_TEXT_OUTPUT_PROTOCOL {
  EFI_TEXT_RESET                  Reset;

  EFI_TEXT_STRING                 OutputString;
  EFI_TEXT_TEST_STRING            TestString;

  EFI_TEXT_QUERY_MODE             QueryMode;
  EFI_TEXT_SET_MODE               SetMode;
  EFI_TEXT_SET_ATTRIBUTE          SetAttribute;

  EFI_TEXT_CLEAR_SCREEN           ClearScreen;
  EFI_TEXT_SET_CURSOR_POSITION    SetCursorPosition;
  EFI_TEXT_ENABLE_CURSOR          EnableCursor;

  ///
  /// Pointer to SIMPLE_TEXT_OUTPUT_MODE data.
  ///
  EFI_SIMPLE_TEXT_OUTPUT_MODE     *Mode;
};
```

These are entirely different protocols from SerialIO. It must be TerminalDxe to install those protocols, as it implements the SimpleText functions of those protocols.

filipleple replied:
Right, thank you for the extra context and detail. I've dropped the `SimpleText*` commit, the adapter works without it

miczyg1 commented on Drivers/OptionRomPkg/Bus/Usb/FtdiUsbSerialDxe/FtdiUsbSerialDriver.c:
```diff
@@ -1390,25 +1410,54 @@ ResetInternal (
+
+  // Now send PURGE_RX (0x1)
+  DevReq.Value       = 0x1; // PURGE_RX
+  Status = UsbSerialDevice->UsbIo->UsbControlTransfer (
```
Why do we send additional USB commands?

filipleple replied:
I saw the SerialIO reset error in debug, so I went and found out that some FTDI chips don't support RESET_PORT_PURGE_RX and require RESET_PORT_RESET (0x0) instead, so I tried to cover all the bases with these extra commands.

Since there's still a SerialIO reset error happening in debug, I'd also think this should be redundant, but no. Without these changes the working console option with the `TerminalTypeTtyTerm` GUID doesn't get listed in BMM for some reason.

filipleple committed:
FtdiUsbSerialDxe.c: don't append redundant FlowControl node

There's no need to append a TtyTerm() node after the Uart node - that is
what TerminalDxe does. Doing so results in a duplicate TtyTerm() node.

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>
